File: cagong_googlemap/lib/models/cafe.dart
import 'package:google_maps_flutter/google_maps_flutter.dart';

/// 카페 모델 - Flask API 응답과 매칭
class Cafe {
  final int id;
  final String name;
  final String address;
  final double latitude;
  final double longitude;
  final String? message;
  final int? hoursWeekday;
  final int? hoursWeekend;
  final String? price;
  final String? videoUrl;
  final String? lastOrder;
  final OperatingHours operatingHours;
  final ReservationInfo reservation;

  Cafe({
    required this.id,
    required this.name,
    required this.address,
    required this.latitude,
    required this.longitude,
    this.message,
    this.hoursWeekday,
    this.hoursWeekend,
    this.price,
    this.videoUrl,
    this.lastOrder,
    required this.operatingHours,
    required this.reservation,
  });

  /// Flask API JSON 응답에서 Cafe 객체 생성
  factory Cafe.fromJson(Map<String, dynamic> json) {
    return Cafe(
      id: json['id'] ?? 0,
      name: json['name'] ?? 'Unnamed Cafe',
      address: json['address'] ?? 'No address',
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      message: json['message'],
      hoursWeekday: json['hours_weekday'],
      hoursWeekend: json['hours_weekend'],
      price: json['price'],
      videoUrl: json['video_url'],
      lastOrder: json['last_order'],
      operatingHours: OperatingHours.fromJson(json['operating_hours'] ?? {}),
      reservation: ReservationInfo.fromJson(json['reservation'] ?? {}),
    );
  }

  /// 구글맵 LatLng 객체로 변환
  LatLng get location => LatLng(latitude, longitude);
}

/// 운영 시간 정보
class OperatingHours {
  final DayHours? monday;
  final DayHours? tuesday;
  final DayHours? wednesday;
  final DayHours? thursday;
  final DayHours? friday;
  final DayHours? saturday;
  final DayHours? sunday;
  final String? description;

  OperatingHours({
    this.monday,
    this.tuesday,
    this.wednesday,
    this.thursday,
    this.friday,
    this.saturday,
    this.sunday,
    this.description,
  });

  factory OperatingHours.fromJson(Map<String, dynamic> json) {
    return OperatingHours(
      monday: json['monday'] != null ? DayHours.fromJson(json['monday']) : null,
      tuesday: json['tuesday'] != null ? DayHours.fromJson(json['tuesday']) : null,
      wednesday: json['wednesday'] != null ? DayHours.fromJson(json['wednesday']) : null,
      thursday: json['thursday'] != null ? DayHours.fromJson(json['thursday']) : null,
      friday: json['friday'] != null ? DayHours.fromJson(json['friday']) : null,
      saturday: json['saturday'] != null ? DayHours.fromJson(json['saturday']) : null,
      sunday: json['sunday'] != null ? DayHours.fromJson(json['sunday']) : null,
      description: json['description'],
    );
  }

  /// 오늘의 운영시간 가져오기
  DayHours? getTodayHours() {
    final now = DateTime.now();
    switch (now.weekday) {
      case DateTime.monday:
        return monday;
      case DateTime.tuesday:
        return tuesday;
      case DateTime.wednesday:
        return wednesday;
      case DateTime.thursday:
        return thursday;
      case DateTime.friday:
        return friday;
      case DateTime.saturday:
        return saturday;
      case DateTime.sunday:
        return sunday;
      default:
        return null;
    }
  }
}

/// 하루 운영 시간
class DayHours {
  final String? begin;
  final String? end;

  DayHours({this.begin, this.end});

  factory DayHours.fromJson(Map<String, dynamic> json) {
    return DayHours(
      begin: json['begin'],
      end: json['end'],
    );
  }

  /// 운영시간 문자열 반환 (예: "09:00 - 22:00")
  String toDisplayString() {
    if (begin == null || end == null) return '정보 없음';
    return '$begin - $end';
  }
}

/// 예약 정보
class ReservationInfo {
  final bool enabled;
  final int? totalSeats;
  final int? totalConsents;
  final String? startTime;
  final String? endTime;
  final int? hourlyRate;

  ReservationInfo({
    required this.enabled,
    this.totalSeats,
    this.totalConsents,
    this.startTime,
    this.endTime,
    this.hourlyRate,
  });

  factory ReservationInfo.fromJson(Map<String, dynamic> json) {
    return ReservationInfo(
      enabled: json['enabled'] ?? false,
      totalSeats: json['total_seats'],
      totalConsents: json['total_consents'],
      startTime: json['start_time'],
      endTime: json['end_time'],
      hourlyRate: json['hourly_rate'],
    );
  }
}

File: cagong_backend/models/cafe.py
from . import db
from datetime import datetime


# 카페 모델인데 예약 관련 필드도 포함

class Cafe(db.Model):
    __tablename__ = 'cafes'
    

    # 기본 정보
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)
    address = db.Column(db.Text, nullable=False)


    # 위치 정보(좌표) - GPS 정밀도를 위해 Numeric 사용
    latitude = db.Column(db.Numeric(precision=10, scale=7), nullable=False)
    longitude = db.Column(db.Numeric(precision=10, scale=7), nullable=False)
    

    # 운영 정보
    message         = db.Column(db.Text)            # 카페 소개 한 줄
    hours_weekday   = db.Column(db.Integer)     # 권장 이용 시간
    hours_weekend   = db.Column(db.Integer)     # 권장 이용 시간
    price           = db.Column(db.String(50))        # 아아 가격
    video_url       = db.Column(db.String(255))   # 카페 내부 영상
    last_order      = db.Column(db.String(50))   # 라스트 오더 시간
    

    # 요일별 운영시간(날마다 다른 경우가 있음...) - HH:MM 형식의 문자열
    monday_begin        = db.Column(db.String(5))      # 예: "09:00"
    monday_end          = db.Column(db.String(5))      # 예: "22:00"
    tuesday_begin       = db.Column(db.String(5))
    tuesday_end         = db.Column(db.String(5))
    wednesday_begin     = db.Column(db.String(5))
    wednesday_end       = db.Column(db.String(5))
    thursday_begin      = db.Column(db.String(5))
    thursday_end        = db.Column(db.String(5))
    friday_begin        = db.Column(db.String(5))
    friday_end          = db.Column(db.String(5))
    saturday_begin      = db.Column(db.String(5))
    saturday_end        = db.Column(db.String(5))
    sunday_begin        = db.Column(db.String(5))
    sunday_end          = db.Column(db.String(5))


    # 추가 정보
    operating_hours = db.Column(db.Text)  # 전체 운영시간 설명(ex 평일 09~20, 주말 14~22 처럼 설명)
    

    # 예약 기본 설정
    reservation_enabled = db.Column(db.Boolean, default=False)  # 예약 기능 활성화 여부
    total_seats = db.Column(db.Integer, default=0)  # 총 좌석 수
    total_consents = db.Column(db.Integer, default=0)  # 총 콘센트 수 (0이면 콘센트 없음)
    

    # 예약 시간 설정 - HH:MM 형식의 문자열
    reservation_start_time = db.Column(db.String(5))  # 예약 시작 시간 (예: "09:00")
    reservation_end_time = db.Column(db.String(5))    # 예약 종료 시간 (예: "22:00")
    

    # 예약 정책
    hourly_rate = db.Column(db.Integer, default=0)     # 30분당 요금(원)


    # 타임스탬프
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)



    def __repr__(self):
        return f'<Cafe {self.name}>'


    def to_dict(self):
        """JSON 응답을 위한 딕셔너리 변환"""
        return {
            'id': self.id,
            'name': self.name,
            'address': self.address,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'message': self.message,
            'hours_weekday': self.hours_weekday,
            'hours_weekend': self.hours_weekend,
            'price': self.price,
            'video_url': self.video_url,
            'last_order': self.last_order,
            'operating_hours': {
                'monday': {
                    'begin': self.monday_begin,
                    'end': self.monday_end
                },
                'tuesday': {
                    'begin': self.tuesday_begin,
                    'end': self.tuesday_end
                },
                'wednesday': {
                    'begin': self.wednesday_begin,
                    'end': self.wednesday_end
                },
                'thursday': {
                    'begin': self.thursday_begin,
                    'end': self.thursday_end
                },
                'friday': {
                    'begin': self.friday_begin,
                    'end': self.friday_end
                },
                'saturday': {
                    'begin': self.saturday_begin,
                    'end': self.saturday_end
                },
                'sunday': {
                    'begin': self.sunday_begin,
                    'end': self.sunday_end
                },
                'description': self.operating_hours
            },
            'reservation': {
                'enabled': self.reservation_enabled,
                'total_seats': self.total_seats,
                'total_consents': self.total_consents,
                'start_time': self.reservation_start_time,
                'end_time': self.reservation_end_time,
                'hourly_rate': self.hourly_rate
            },
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
